src/util/shmem/fd_shmem_user.c

```diff
@@ -74,6 +75,59 @@ fd_shmem_private_map_query_by_addr( fd_shmem_join_info_t * map,
   return def;
 }
 
+/*
+ * fd_shmem_private_grab_region will attempt to map a region at the passed 
+ * address with the passed size. If the return value of `mmap` equals the 
+ * passed address this means the area of memory was unmapped previously and
+ * we have succesfully "grabbed" the region. We can then call `mmap` with 
+ * MAP_FIXED over the region and be certain no corruption occurs. If the 
+ * return value of `mmap` does not return the passed address this means that 
+ * the passed region is already atleast partially mapped and we cannot grab it.
+ */
+static int
+fd_shmem_private_grab_region( ulong addr,
+                              ulong size ) {
+  void *mmap_ret;
+
+  mmap_ret = mmap( (void*)addr, size, PROT_READ, MAP_ANON|MAP_PRIVATE, -1, 0 );
+
+  if( mmap_ret == MAP_FAILED )
+    return 0;
+
+  /* Only call munmap on failure case. On success we want to keep the mapping */
+  if( (ulong)mmap_ret != addr ) {
+    munmap( mmap_ret, size );
+    return 0;
+  }
+
+  return 1;
+}
+
+static ulong
+fd_shmem_private_get_random_mappable_addr( ulong size, 
+                                           ulong page_size ) {
+  ulong ret_addr = 0;
+
+  for(;;){
```

Comment:
I would do `for( ulong i=0; i<1000; i++ ) {` and `FD_LOG_ERR(( "unable to find random address for memory map" ))` or whatever if you get past the loop so we can't hang here. Maybe use a different value than 1000, depends how often this loop will fail but you want to make it 0.00000001% or whatever.